### PR TITLE
Keycloak OIDC Provider

### PR DESCRIPTION
**Pull request by NickMeves**
Make Keycloak Provider that completely uses OIDC functionality

## Description

Keycloak is OIDC and we should deprecate the OAuth based `keycloak` provider that supports limited functionality.

## Motivation and Context

We get all the feature sets we add to the OIDCProvider. Any Keycloak specific extensions (e.g. Roles from Access Token claims) can be layered on top of OIDC capabilities.

## How Has This Been Tested?

Unit Tests - Keycloak community testing & feedback desired.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: providers/kaycloak_oidc.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package providers
```
Comment (JoelSpeed) [inline providers/kaycloak_oidc.go:1]: There's a typo in the filename here

File: providers/kaycloak_oidc.go [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+
+const keycloakOIDCProviderName = "Keycloak OIDC"
+
+// KeycloakOIDCProvider creates a Keycloak provider based on OIDCProvider
+type KeycloakOIDCProvider struct {
+	*OIDCProvider
+}
```
Comment (JoelSpeed) [inline providers/kaycloak_oidc.go:8]: So this is literally just the OIDC provider? What's the benefit of having this, where are we going?

Comment (NickMeves) [reply providers/kaycloak_oidc.go:8]: People are asking about the Roles from Access Token feature. I whipped this together as a starting point for them.

`EnrichSession` would likely be overridden to extract Roles from the Access Token claims like this PR started: https://github.com/oauth2-proxy/oauth2-proxy/pull/767/files

Comment (NickMeves) [reply providers/kaycloak_oidc.go:8]: https://github.com/oauth2-proxy/oauth2-proxy/issues/976#issuecomment-798931991

Comment (NickMeves) [reply providers/kaycloak_oidc.go]: I updated the dummy PR to show the others exactly where to plugin the role logic extension they wanted 👍

**Comment by NickMeves** [comment]
@khawaga I pulled in some of your Access Tokens claims logic from #767 here.

@khawaga or @tpohl Can one of you run with this and get it over the finish line in your own PR? Access Token logic needs unit tests and manual integration tests making sure everything with Keycloak + OIDC is OK.

Since I don't use Keycloak myself, I'm gonna pass the baton off at this point 😄

**Comment by tpohl** [comment]
@NickMeves Thank you for this quick action! I can confirm that on first sight this is working nicely with our Keycloak Integration (I have tested my original requirement to allow or deny based upon a client role).

I am going to test further on our development environment.

**Comment by NickMeves** [comment]
@tpohl Thanks for taking a look! I'm glad it actually worked 😅  -- I didn't even test that it compiled.

Some cases I'm curious about for live tests + unit test cases:
- normal roles
- client based roles of the form `clientName:roleName`
- No role support enabled in Keycloak or set as restriction in OAuth2 Proxy
- Roles setup in Keycloak, but no `keycloak-role` config set (So we'd still see `role:roleName` in the `X-Forwarded-Groups`)

This won't be going into v7.1 we are focusing on, but I'll circle back and checkin on how test cases are looking if one of you picks them up before we roll out v7.2.

**Comment by tpohl** [comment]
Yes, it does compile ;).

I have a small testsetup with a httpbin service behind the proxy and it actually works really well. This is what comes through (reduced a bit, but still ;)).

I will try to figure out, how to do the unit tests and see, if I can setup a complete Environment to test some edge cases.

`{
  "args": {},
  "headers": {
    "Accept": "application/json",
    "Accept-Encoding": "gzip, deflate",
    "Accept-Language": "de-de",
    "Cookie": "_oauth2_proxy_0=xxxxxxx",
    "Host": "localhost:4180",
    "Referer": "http://localhost:4180/",
    "User-Agent": "xxx",
    "X-Forwarded-Email": "tpohl@github.com",
    "X-Forwarded-Groups": "role:a_realm_role,role:a-client:a-client-role",
    "X-Forwarded-Preferred-Username": "tpohl@github.com",
    "X-Forwarded-User": "d5443a7a-85a5-11eb-8dcd-0242ac130003"
  },
  "origin": "172.20.0.1",
  "url": "http://localhost:4180/get"
}`

**Comment by pb82** [comment]
@NickMeves thanks for the PR, that is really useful. I can try to add the missing unit tests and test it once more. Hope we can get this over the line soon.

**Comment by NickMeves** [comment]
> @NickMeves thanks for the PR, that is really useful. I can try to add the missing unit tests and test it once more. Hope we can get this over the line soon.

Thanks, I appreciate the help! I started at a new company yesterday so I'll be busy spinning up for the next few weeks and won't have much time for OAuth2 Proxy.

**Comment by pb82** [comment]
@NickMeves I've rebased your PR against the latest changes on master. The options structure has changed quite a bit. Do you want me to force push into your branch or would you prefer a new PR?

File: providers/keycloak_oidc.go [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+		OIDCProvider: &OIDCProvider{
+			ProviderData: p,
+		},
+	}
+}
+
+var _ Provider = (*KeycloakOIDCProvider)(nil)
+
+// AddAllowedRoles sets Keycloak roles that are authorized.
+// Assumes `SetAllowedGroups` is already called on groups and appends to that
```
Comment (pb82) [inline providers/keycloak_oidc.go:31]: @tpohl is there a scenario where only roles, but no groups are in the allowed list? In that case this results in a NPE. I'll add a check for that.

Comment (tpohl) [reply providers/keycloak_oidc.go:31]: yes, I would not make an assumption, that an allowedGroup is set

**Comment by NickMeves** [comment]
> @NickMeves I've rebased your PR against the latest changes on master. The options structure has changed quite a bit. Do you want me to force push into your branch or would you prefer a new PR?

Sorry for the delay! Whatever is easiest for you, go ahead! Still concentrating on spinning up at the new job, so I might be missing for sporadic stetches and not pay attention to this.

**Comment by pb82** [comment]
@NickMeves No worries, thanks! I'll create a new PR that includes yours and my commits and reference this one. Best of luck for your new venture!

**Comment by NickMeves** [comment]
> @NickMeves No worries, thanks! I'll create a new PR that includes yours and my commits and reference this one. Best of luck for your new venture!

Thanks! For reference, we are trying to write all new tests in Ginkgo - here's a reference from the `providers` package: https://github.com/oauth2-proxy/oauth2-proxy/blob/06808704a3762de73bed1cc92b077420eb93d735/providers/keycloak_test.go#L48
